### DISCUSSION TITLE
Does ZoneTree support multi threads writing to the same database?

### DISCUSSION THREAD

**Discussion by aley1** [koculu/ZoneTree#98] [Q&A] [answered]
Does ZoneTree support multi threads writing to the same database?

e.g. Create single database instance and share it among threads in a thread pool. Threads can read or write to the database.

**Comment by koculu** [comment, marked as answer]
Yes, it does.
ZoneTree synchronizes the writes and reads internally and you don't have to worry about synchronization at all.
The following read/write methods are thread-safe:
ContainsKey, Upsert, TryGet, TryAdd, TryDelete, ForceDelete, Count

Please also consider the following in multi-threaded use-cases:
- Thread safety does not guarantee data consistency.
- Data consistency does not imply thread safety.
